runners/direct-java/src/main/java/org/apache/beam/runners/direct/BoundedReadEvaluatorFactory.java

```diff
@@ -83,7 +83,8 @@ public void cleanup() {}
       if (sourceEvaluators.putIfAbsent(transform, evaluatorQueue) == null) {
         // If no queue existed in the evaluators, add an evaluator to initialize the evaluator
         // factory for this transform
-        BoundedSource<OutputT> source = transform.getTransform().getSource();
+        Bounded<OutputT> bound = (Bounded<OutputT>) transform.getTransform();
```

Comment:
We can add necessary configuration to support IDE's, but we shouldn't make negative changes to the code base. In this case, the unsafe-cast moves validation from compile-time to runtime.

For this error, it appears that we can instead change the type signature for `AppliedPTransform` to `<InputT extends PInput, OutputT extends POutput, TransformT extends PTransform<InputT, OutputT>>` (drop the wildcard on the application of `InputT`)